{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###    Pipeline to illustrate GWAS ina tetraploid species\n",
    "    It uses gen format instead of vcf. gen format contains\n",
    "       chr pos alleles ....\n",
    "    space / tab separated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import gzip\n",
    "import sys\n",
    "import time\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# specific seqbreed modules\n",
    "import genome as gg\n",
    "import selection as sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# current dir\n",
    "cdir = os.getcwd()\n",
    "\n",
    "# working directory\n",
    "wdir = cdir\n",
    "\n",
    "# input file dir\n",
    "ddir = cdir\n",
    "\n",
    "# working files, seqfile contains snp positions from vcf\n",
    "# this file is generated by gg.GFounder below\n",
    "seqfile = 'seq.pos'\n",
    "\n",
    "# list of chips that can be analyzed, include sequence as output from gg.GFounder()\n",
    "chipfiles = [seqfile]\n",
    "# contains qtn positions\n",
    "qtnfile = ddir + '/potato.qtn.pos'\n",
    "# gene dropping pedigree\n",
    "pedfile = ddir + '/potato.ped'\n",
    "# snp file in gen format\n",
    "genfile = ddir + '/potato.gen.gz'\n",
    "\n",
    "# goto working directory\n",
    "os.chdir(wdir)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP 1:\n",
    "# uploads genotypes and generates snp positions (snpFile)\n",
    "# NOTE that ploidy level must be specified if gen format\n",
    "gbase = gg.GFounder(vcfFile=genfile, snpFile=seqfile, ploidy=4)\n",
    "\n",
    "# STEP 2: generates Genome object with chr names, recombination map, etc\n",
    "gfeatures = gg.Genome(snpFile=seqfile, ploidy=gbase.ploidy)\n",
    "\n",
    "# prints some basic info\n",
    "gfeatures.print()\n",
    "\n",
    "# STEP 3: read QTN file\n",
    "qtn = gg.QTNs(h2=[0.5, 0.7], genome=gfeatures, qtnFile=qtnfile)\n",
    "qtn.get_var(gfeatures,gbase)\n",
    "\n",
    "# STEP 4: generate base population\n",
    "pop = gg.Population(gfeatures, pedFile=None, generation=None, qtns=qtn, gfounders=gbase)\n",
    "qtn.print(gfeatures)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP 5: Simulate a F2 between the 10 most extreme lowest and highest lines for second phenotype\n",
    "#--> Sort phenotypes and sel parents\n",
    "itrait = 1\n",
    "nsel = 10\n",
    "y = np.array(list(pop.inds[i].y[itrait] for i in range(pop.n))) # phenotypes\n",
    "yindex = np.argsort(y)\n",
    "ilow = yindex[:nsel]\n",
    "ihigh = np.random.permutation(yindex[-nsel:]) # random mating between extremes\n",
    "\n",
    "#--> F1 between lowest and highest inds\n",
    "n0 = pop.n\n",
    "for i in range(nsel):\n",
    "    parents = [pop.inds[ilow[i]], pop.inds[ihigh[i]] ]\n",
    "    pop.addInd(parents, genome=gfeatures, gfounders=gbase, qtns=qtn, id=None, sex=None, t=1)\n",
    "\n",
    "#--> F2 by selfcrossing\n",
    "noff = 20\n",
    "n1 = pop.n\n",
    "for i in range(nsel):\n",
    "    iid = n0+i # index of parent\n",
    "    parents = [pop.inds[iid], pop.inds[iid]]\n",
    "    for j in range(noff):\n",
    "        pop.addInd(parents, genome=gfeatures, gfounders=gbase, qtns=qtn, id=None, sex=None, t=2)\n",
    "\n",
    "# STEP 5: GWAS\n",
    "# generate chip objects\n",
    "chipseq = gg.Chip(chipFile=seqfile, genome=gfeatures, name='seq_chip')\n",
    "\n",
    "# get genotypes\n",
    "X = gg.do_X(pop.inds, gfeatures, gbase, chip=chipseq)\n",
    "\n",
    "# GWAS for second trait\n",
    "gwas = sel.Gwas(X, chipseq)\n",
    "gwas.fit(pop.inds, trait=1)\n",
    "gwas.plot(fdr=True)    # FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GWAS for first trait\n",
    "gwas = sel.Gwas(X, chipseq)\n",
    "gwas.fit(pop.inds, trait=0)\n",
    "gwas.plot(fdr=True)    # FDR\n",
    "gwas.print(gfeatures)  # prints gwas results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appendix: How to simulate repeated measurements\n",
    "itrait = 0\n",
    "current_y = np.array(list(ind.y[itrait] for ind in pop.inds))\n",
    "new_y = np.array([])\n",
    "for ind in pop.inds:\n",
    "    y = ind.g_add[itrait] + ind.g_dom[itrait] + np.random.normal(0, qtn.se[itrait])\n",
    "    new_y = np.append(new_y, y)\n",
    "\n",
    "# plot new_y vs current_y\n",
    "plt.scatter(new_y, current_y)\n",
    "plt.title('Repeated measurements: New vs Old phenotypes')\n",
    "plt.xlabel('New')\n",
    "plt.ylabel('Old')\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
